cranelift/peepmatic/crates/runtime/src/integer_interner.rs

```diff
@@ -0,0 +1,71 @@
+//! Interner for (potentially large) integer values.
+//!
+//! We support matching on integers that can be represented by `u64`, but only
+//! support automata results that fit in a `u32`. So we intern the (relatively
+//! few compared to the full range of `u64`) integers we are matching against
+//! here and then reference them by `IntegerId`.
+
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::convert::TryInto;
+
+/// An identifier for an interned integer.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct IntegerId(#[doc(hidden)] pub u32);
+
+/// An interner for integer values.
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct IntegerInterner {
+    // Note: we use `BTreeMap`s for deterministic serialization.
+    map: BTreeMap<u64, IntegerId>,
```

Comment:
Maybe construct this map on the fly from `values` when deserializing?

Reply:
This is something we could do in the future with a custom serialization and deserialization implementation.